File: С++/Algorithms/Medium/1. Deepest Leaves Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int deepestLeavesSum(TreeNode* root) {

        stack<pair<TreeNode *, int>> nodes;
        nodes.push(make_pair(root, 0));
        unordered_map<int, int> dict;
        
        int sum = 0, deep = 0;
        while (!nodes.empty()) {
            root = nodes.top().first;
            int val = nodes.top().second;
            nodes.pop();
            
            if (root->left)
                nodes.push(make_pair(root->left, val + 1));
            
            if (root->right)
                nodes.push(make_pair(root->right, val + 1));
            
            dict[val] += root->val;
            
            if (val > deep) {
                deep = val;
            }
            
        }
                   
        return dict[deep];
    }
};

File: Java/Algorithms/Medium/1. Deepest Leaves Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    static int max = -1;
    static int sum = 0;
    public static void func(TreeNode root, int level){
        if(root == null) return;
        if(level > max){
            max = level;
            sum = root.val;
        } else if(max == level) sum += root.val;
        func(root.left, level + 1);
        func(root.right, level + 1);
    }
    public int deepestLeavesSum(TreeNode root) {
        sum = 0;
        max = 0;
        func(root, 0);
        return sum;
    }
}
